File: easycontrol/server/src/main/java/top/saymzx/easycontrol/server/entity/PointersState.java
/*
 * 本项目大量借鉴学习了开源投屏软件：Scrcpy，在此对该项目表示感谢
 */
package top.saymzx.easycontrol.server.entity;

import android.view.MotionEvent;

import java.util.concurrent.ConcurrentHashMap;

public final class PointersState {

  private final int MAX_POINTERS = 10;

  private final ConcurrentHashMap<Integer, Pointer> pointers = new ConcurrentHashMap<>();
  public final MotionEvent.PointerProperties[] pointerProperties = new MotionEvent.PointerProperties[MAX_POINTERS];
  public final MotionEvent.PointerCoords[] pointerCoords = new MotionEvent.PointerCoords[MAX_POINTERS];

  public PointersState() {
    // 初始化指针
    for (int i = 0; i < MAX_POINTERS; ++i) {
      MotionEvent.PointerProperties props = new MotionEvent.PointerProperties();
      props.toolType = MotionEvent.TOOL_TYPE_FINGER;
      pointerProperties[i] = props;

      MotionEvent.PointerCoords coords = new MotionEvent.PointerCoords();
      coords.orientation = 0;
      coords.size = 0.01f;
      coords.pressure = 1f;
      pointerCoords[i] = coords;
    }
  }

  public Pointer newPointer(int pointerId, long now) {
    for (int i = 0; i < MAX_POINTERS; i++) {
      if (isLocalIdAvailable(i)) {
        Pointer pointer = new Pointer(i, now);
        pointers.put(pointerId, pointer);
        return pointer;
      }
    }
    return null;
  }

  public Pointer get(int pointerId) {
    return pointers.get(pointerId);
  }

  private boolean isLocalIdAvailable(int localId) {
    for (Pointer value : pointers.values()) if (value.id == localId) return false;
    return true;
  }

  public void remove(int pointerId) {
    pointers.remove(pointerId);
  }

  public int update() {
    int i = 0;
    for (Pointer value : pointers.values()) {
      pointerProperties[i].id = value.id;
      pointerCoords[i].x = value.x;
      pointerCoords[i].y = value.y;
      i++;
    }
    return i;
  }

}

File: scrcpy_android/server/src/main/java/top/saymzx/scrcpy/server/entity/PointersState.kt
package top.saymzx.scrcpy.server.entity

import android.view.MotionEvent.PointerCoords
import android.view.MotionEvent.PointerProperties

class PointersState {
  private val pointers: MutableList<Pointer> = ArrayList()
  private fun indexOf(id: Int): Int {
    for (i in pointers.indices) {
      val pointer = pointers[i]
      if (pointer.id == id) {
        return i
      }
    }
    return -1
  }

  private fun isLocalIdAvailable(localId: Int): Boolean {
    for (i in pointers.indices) {
      val pointer = pointers[i]
      if (pointer.localId == localId) {
        return false
      }
    }
    return true
  }

  private fun nextUnusedLocalId(): Int {
    for (localId in 0 until MAX_POINTERS) {
      if (isLocalIdAvailable(localId)) {
        return localId
      }
    }
    return -1
  }

  operator fun get(index: Int): Pointer {
    return pointers[index]
  }

  fun getPointerIndex(id: Int): Int {
    val index = indexOf(id)
    if (index != -1) {
      // already exists, return it
      return index
    }
    if (pointers.size >= MAX_POINTERS) {
      // it's full
      return -1
    }
    // id 0 is reserved for mouse events
    val localId = nextUnusedLocalId()
    if (localId == -1) {
      throw AssertionError("pointers.size() < maxFingers implies that a local id is available")
    }
    val pointer = Pointer(id, localId)
    pointers.add(pointer)
    // return the index of the pointer
    return pointers.size - 1
  }

  /**
   * Initialize the motion event parameters.
   *
   * @param props  the pointer properties
   * @param coords the pointer coordinates
   * @return The number of items initialized (the number of pointers).
   */
  fun update(props: Array<PointerProperties>, coords: Array<PointerCoords>): Int {
    val count = pointers.size
    for (i in 0 until count) {
      val pointer = pointers[i]

      // id 0 is reserved for mouse events
      props[i].id = pointer.localId
      coords[i].x = pointer.x
      coords[i].y = pointer.y
      coords[i].pressure = pointer.pressure
    }
    cleanUp()
    return count
  }

  /**
   * Remove all pointers which are UP.
   */
  private fun cleanUp() {
    for (i in pointers.indices.reversed()) {
      val pointer = pointers[i]
      if (pointer.isUp) {
        pointers.removeAt(i)
      }
    }
  }

  companion object {
    const val MAX_POINTERS = 10
  }
}